File: facebook/src/com/facebook/internal/UrlRedirectCache.java
/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.internal;

import android.net.Uri;
import android.util.Log;
import com.facebook.LoggingBehavior;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;

/**
 * com.facebook.internal is solely for the use of other packages within the Facebook SDK for
 * Android. Use of any of the classes in this package is unsupported, and they may be modified or
 * removed without warning at any time.
 */
class UrlRedirectCache {
    static final String TAG = UrlRedirectCache.class.getSimpleName();
    private static final String REDIRECT_CONTENT_TAG = TAG + "_Redirect";

    private volatile static FileLruCache urlRedirectCache;

    synchronized static FileLruCache getCache() throws IOException{
        if (urlRedirectCache == null) {
            urlRedirectCache = new FileLruCache(TAG, new FileLruCache.Limits());
        }
        return urlRedirectCache;
    }

    static Uri getRedirectedUri(Uri uri) {
        if (uri == null) {
            return null;
        }

        String uriString = uri.toString();
        InputStreamReader reader = null;
        try {
            InputStream stream;
            FileLruCache cache = getCache();
            boolean redirectExists = false;
            while ((stream = cache.get(uriString, REDIRECT_CONTENT_TAG)) != null) {
                redirectExists = true;

                // Get the redirected url
                reader = new InputStreamReader(stream);
                char[] buffer = new char[128];
                int bufferLength;
                StringBuilder urlBuilder = new StringBuilder();
                while ((bufferLength = reader.read(buffer, 0, buffer.length)) > 0) {
                    urlBuilder.append(buffer, 0, bufferLength);
                }
                Utility.closeQuietly(reader);

                // Iterate to the next url in the redirection
                uriString = urlBuilder.toString();
            }

            if (redirectExists) {
                return Uri.parse(uriString);
            }
        } catch (IOException ioe) {
        } finally {
            Utility.closeQuietly(reader);
        }

        return null;
    }

    static void cacheUriRedirect(Uri fromUri, Uri toUri) {
        if (fromUri == null || toUri == null) {
            return;
        }

        OutputStream redirectStream = null;
        try {
            FileLruCache cache = getCache();
            redirectStream = cache.openPutStream(fromUri.toString(), REDIRECT_CONTENT_TAG);
            redirectStream.write(toUri.toString().getBytes());
        } catch (IOException e) {
            // Caching is best effort
        } finally {
            Utility.closeQuietly(redirectStream);
        }
    }

    static void clearCache() {
        try {
            getCache().clearCache();
        } catch (IOException e) {
            Logger.log(LoggingBehavior.CACHE, Log.WARN, TAG, "clearCache failed " + e.getMessage());
        }
    }
}

File: facebook-core/src/main/java/com/facebook/internal/UrlRedirectCache.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.internal

import android.net.Uri
import android.util.Log
import com.facebook.LoggingBehavior
import java.io.IOException
import java.io.InputStreamReader
import java.io.OutputStream
import java.util.HashSet

/**
 * com.facebook.internal is solely for the use of other packages within the Facebook SDK for
 * Android. Use of any of the classes in this package is unsupported, and they may be modified or
 * removed without warning at any time.
 */
object UrlRedirectCache {
  private val tag = UrlRedirectCache::class.simpleName ?: "UrlRedirectCache"
  private val redirectContentTag = tag + "_Redirect"
  private var urlRedirectFileLruCache: FileLruCache? = null

  @Synchronized
  @JvmStatic
  @Throws(IOException::class)
  fun getCache(): FileLruCache {
    val nonNullCache = urlRedirectFileLruCache ?: FileLruCache(tag, FileLruCache.Limits())
    urlRedirectFileLruCache = nonNullCache
    return nonNullCache
  }

  @JvmStatic
  fun getRedirectedUri(uri: Uri?): Uri? {
    if (uri == null) {
      return null
    }

    var uriString = uri.toString()
    var reader: InputStreamReader? = null
    val redirectChain = HashSet<String>()
    redirectChain.add(uriString)
    try {
      val cache = getCache()
      var redirectExists = false
      var stream = cache.get(uriString, this.redirectContentTag)

      while (stream != null) {
        redirectExists = true
        // Get the redirected url
        reader = InputStreamReader(stream)
        val buffer = CharArray(128)
        val urlBuilder = StringBuilder()
        var bufferLength = reader.read(buffer, 0, buffer.size)
        while (bufferLength > 0) {
          urlBuilder.append(buffer, 0, bufferLength)
          bufferLength = reader.read(buffer, 0, buffer.size)
        }
        Utility.closeQuietly(reader)

        // Iterate to the next url in the redirection
        val redirectToUriString = urlBuilder.toString()
        if (redirectChain.contains(redirectToUriString)) {
          if (redirectToUriString == uriString) {
            // uriString redirect to itself. Stop the loop
            break
          } else {
            // A loop with more than 1 address is detected. It's unexpected.
            // In this case, return null so that the caller can directly use the original address.
            Logger.log(
                LoggingBehavior.CACHE, Log.ERROR, this.tag, "A loop detected in UrlRedirectCache")
            return null
          }
        }
        uriString = redirectToUriString
        redirectChain.add(uriString)
        stream = cache.get(uriString, this.redirectContentTag)
      }

      if (redirectExists) {
        return Uri.parse(uriString)
      }
    } catch (e: IOException) {
      Logger.log(
          LoggingBehavior.CACHE,
          Log.INFO,
          this.tag,
          "IOException when accessing cache: " + e.message)
    } finally {
      Utility.closeQuietly(reader)
    }

    return null
  }

  @JvmStatic
  fun cacheUriRedirect(fromUri: Uri?, toUri: Uri?) {
    if (fromUri == null || toUri == null) {
      return
    }

    var redirectStream: OutputStream? = null
    try {
      val cache = getCache()
      redirectStream = cache.openPutStream(fromUri.toString(), this.redirectContentTag)
      redirectStream.write(toUri.toString().toByteArray())
    } catch (e: IOException) {
      // Caching is best effort
      Logger.log(
          LoggingBehavior.CACHE,
          Log.INFO,
          this.tag,
          "IOException when accessing cache: " + e.message)
    } finally {
      Utility.closeQuietly(redirectStream)
    }
  }

  @JvmStatic
  fun clearCache() {
    try {
      getCache().clearCache()
    } catch (e: IOException) {
      Logger.log(LoggingBehavior.CACHE, Log.WARN, tag, "clearCache failed " + e.message)
    }
  }
}
